src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -200,9 +202,53 @@ private void AddCompilationRoots()
                     AddCompilationRootsForMainMethod(module);
 
                 AddCompilationRootsForRuntimeExports(module);
-           }
+            }
+
+            if (_options.MultiFile && StartupCodeMain == null)
+            {
+                foreach (var inputFile in _typeSystemContext.InputFilePaths)
+                {
+                    var module = _typeSystemContext.GetModuleFromPath(inputFile.Value);
+
+                    foreach (TypeDesc type in module.GetAllTypes())
+                    {
+                        if (!(type is EcmaType))
+                            continue;
+
+                        // Delegates have empty method IL bodies for the invoke methods. Skip for now.
+                        if (type.IsDelegate || type.ContainsGenericVariables)
+                            continue;
+
+                        EcmaType ecmaType = type as EcmaType;
+
+                        if (ecmaType.Attributes.HasFlag(System.Reflection.TypeAttributes.Public))
+                        {
+                            foreach (var m in ecmaType.GetMethods())
+                            {
+                                if (!(m is EcmaMethod))
+                                    continue;
+
+                                if (m.IsIntrinsic || m.ContainsGenericVariables)
+                                    continue;
+
+                                EcmaMethod ecmaMethod = m as EcmaMethod;
+
+                                if (ecmaMethod.ImplAttributes.HasFlag(System.Reflection.MethodImplAttributes.InternalCall))
+                                    continue;
+
+                                if (ecmaMethod.Attributes.HasFlag(System.Reflection.MethodAttributes.Public) &&
```

Comment:
It would probably work better to include all methods for now because of generics and inlining (we will worry about version resilient multifile codegen later).